### DISCUSSION TITLE
Idiomtic multi-phase modifiers

### DISCUSSION THREAD

**Discussion by axyz** [floating-ui/floating-ui#1305] [Q&A] [answered]
Hi,
as part of a component library I want to provide certain custom modifiers.

Very often modifiers need to operate on different phases and I am wondering what is the best way to do that without sacrificing performance and keeping a clean API.

These are some of the options I could think of:

### 1. provide multiple modifiers for a single task
```tsx
{ modifiers: [ myModifierRead, myModifierMain, myModifierWrite] }
```
This seems to be the approach used internally in the modifiers bundled with popper: most of them read some stuff from the DOM, others do pure calculations on main and then a single applyStyles modifier write the calculated styles from many different modifiers in the DOM.

doubts on this approach:
1. the API is not great: consumers need to have a different named import for each phase and be careful to include all the phases required for that modifier. This could be improved with a factory function or exporting an array, but seems not very idiomatic to import and use modifiers in different ways depending on how it is authored
2. is there a performance impact on having many different modifiers in the write phase instead of a single one that uses all the modifiers data available?

### 2. do all the logic in `write` or `beforeWrite` phase
this allow to have a single export in the expected modifier format and have complete decoupling between modifiers. The needed calculations and the setting of the style props is all done in the same function that has to be in a late phase.

1. Is there a performance impact on having some calculations on the write phases?
2. What happens if the modifier wants to modify some data of other modifiers that will be used in the write phase as well (e.g. a custom modifier does some calculations in `beforeWrite`, then tries to modify the rects or the offset in the modifiers data and then modifies the styles. Will it have to set `reset` to `true` for the other modifiers to use the changed data? Does it result on additional avoidable cycles?)

### 3. many small modifiers in the main phase and single modifiers in read and write phases that operate on the DOM based on what they find in the modifiers data
```tsx
{ modifiers: [theBigReadModifier, modifier1, modifier2, modifier3, theBigWriteModifier] }

// modifier1

.. do stuff in main and save data in modifiersData.modifier1

// theBigWriteModifier
..
if (state.modifiersData.modifier1) {
  write relevant stuff
}
```

This could be a viable approach if the modifiers live in a single codebase and some conventions are used for the read and write phases to be able to know what to do based on the used modifiers.

1. Is there a performance benefit in doing so?
2. how to handle dependencies? `theBigWriteModifier` depends on modifiersData from modifiers that may or may not be used and relies on checking if the data object of a specific modifier is available or not. On the other hand all the main phase modifier do not have dependencies on the read or write phase modifiers, but they will not do anything useful without them.

---

In general this seem a bit confusing and was wondering if you have suggestions on what could be the most idiomatic and performant approach to provide a small libraries of opt-in modifiers. I would like to avoid to have certain modifiers included by default as this prevent possible treeshaking optimizations: certain components may only need 1 or 2 modifiers and do not have to bundle the rest.

Thanks

**Comment by FezVrasta** [comment, marked as answer]
> 1. the API is not great: consumers need to have a different named import for each phase and be careful to include all the phases required for that modifier. This could be improved with a factory function or exporting an array, but seems not very idiomatic to import and use modifiers in different ways depending on how it is authored

👋  I think the cleanest way to import multiple modifiers is to have them exported as a single array that can then be used as follows:

```
createPopper({
  modifiers: [
    ...customBundleOfModifiers,
  ]
});
```

> 2. is there a performance impact on having many different modifiers in the write phase instead of a single one that uses all the modifiers data available?

The reason why modifiers are divided by phase is to avoid the situation where the library reads and writes to the DOM in random order, the `main` phase is meant as a way to make sure you don't end up doing a lot of write-or-read -> logic -> write-or-read -> logic, etc. On top of that, having all the computations in a middle phase allow them to get access to anything read by the read-modifiers, and allow the write-modifiers to have access to the final data.

What I found to produce the best results (perf wise) is to read the DOM all at once, and then write back to it all at once again.

> ### 2. do all the logic in `write` or `beforeWrite` phase
> 1. Is there a performance impact on having some calculations on the write phases?

This really depends by how heavy the calculations are, I wouldn't worry too much to be honest, you can easily run a Chrome Dev Tools performance capture and take a look at the performance of your modifier.

> 2. What happens if the modifier wants to modify some data of other modifiers that will be used in the write phase as well (e.g. a custom modifier does some calculations in `beforeWrite`, then tries to modify the rects or the offset in the modifiers data and then modifies the styles. Will it have to set `reset` to `true` for the other modifiers to use the changed data? Does it result on additional avoidable cycles?)

Reset is a way to restart the cycle in case some data that previous modifiers used has been changed by a later modifier. The most common use case is when a modifier changes the computed placement value.

> 1. Is there a performance benefit in doing so?

It really depends by how your modifiers logic runs, you should do a perf tools investigation.

> 2. how to handle dependencies? `theBigWriteModifier` depends on modifiersData from modifiers that may or may not be used and relies on checking if the data object of a specific modifier is available or not. On the other hand all the main phase modifier do not have dependencies on the read or write phase modifiers, but they will not do anything useful without them.

Maybe in your case it would make sense to export a function that will return an array of modifiers based on some options passed to it?

**Reply by axyz** [reply]
> I think the cleanest way to import multiple modifiers is to have them exported as a single array that can then be used as follows:

The main problem with that approach is how to pass options.
this is what I usually do:
```tsx
modifiers: [
  { ...customModifer, enabled: someBoolean, options: { padding: someValue } }
]
```

but if we return an array of modifiers we do not have an easy way to pass options to them
```tsx
{ ...customModifer[0], enabled: someBoolean, options: { padding: someValue }, ...customModifier.splice(1) } // this is not nice
```

something like a function to generate that array may work like:
```tsx
modifiers: { ...myModifier(options) }
```
but that can generate confusion and require extra documentation if published and advertised as a popper modifier as consumer would expect a `Modifier`object to be the export

**Reply by FezVrasta** [reply]
I think `...myModifier(options)` could work. I see how it's not the most ergonomic of the solutions but the Popper's modifiers manager already weights quite a lot so I didn't want to make it even heavier by adding more functionalities to it.

I see how an ideal solution would be to allow modifiers to be defined as:

```js
{
  name: 'multi-phase-modifier',
  fn: {
    main: mainLogicFn,
    write: writeLogicFn,
  },
  options: {

  }
}
```

and I may even consider supporting this interface in the next major version.

**Reply by axyz** [reply]
Related: https://github.com/popperjs/popper-core/issues/1306
